internal/resource.go

```diff
@@ -1,140 +1,265 @@
-	err = requests.
-		URL(u).
+
+	req := requests.URL(u).
 		Header("Accept", "*/*").
-		ToFile(fileName).
-		Fetch(ctx)
+		ToFile(fileName)
+
+	err := req.Fetch(ctx)
 	if err != nil {
-		return "", fmt.Errorf("failed to download '%s': %s", u, err)
+		return "", fmt.Errorf("failed to download '%s': %v", u, err)
 	}
-	log.Debug().Str("URL", u).Msg("Downloaded")
+	log.Debug().Msg("Download completed")
 	return fileName, nil
 }
 
-// GetUrlToDir downloads the given resource to the given directory and returns the path to it.
 func GetUrlToDir(u string, targetDir string, ctx context.Context) (string, error) {
-	// create temporary name in the target directory.
 	h := sha256.New()
 	h.Write([]byte(u))
 	fileName := filepath.Join(targetDir, fmt.Sprintf(".%s", hex.EncodeToString(h.Sum(nil))))
 	return getUrl(u, fileName, ctx)
 }
 
-// GetUrlWithDir downloads the given resource to a temporary file and returns the path to it.
 func GetUrltoTempFile(u string, ctx context.Context) (string, error) {
 	file, err := os.CreateTemp("", "prefix")
 	if err != nil {
-		log.Fatal().Err(err)
+		return "", err
 	}
-	fileName := file.Name()
-	return getUrl(u, fileName, ctx)
+	return getUrl(u, file.Name(), ctx)
 }
 
-func (l *Resource) Download(dir string, mode os.FileMode, ctx context.Context) error {
-	ok := false
-	algo, err := getAlgoFromIntegrity(l.Integrity)
+func GetFileHash(filePath string) (string, error) {
+	data, err := ioutil.ReadFile(filePath)
 	if err != nil {
+		return "", fmt.Errorf("failed to read file: %w", err)
+	}
+
+	hasher := sha256.New()
+	hasher.Write(data)
+	return hex.EncodeToString(hasher.Sum(nil)), nil
+}
+
+func (r *Resource) Contains(url string) bool {
+	for _, u := range r.Urls {
+		if u == url {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *Resource) Download(dir string, mode os.FileMode, ctx context.Context) error {
+	if r.CacheUri != "" {
+		token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+		if token == "" {
+			log.Debug().Msg("GRABIT_ARTIFACTORY_TOKEN not set, skipping cache")
+		} else {
+			log.Debug().Msg("Attempting cache download")
+			if err := r.downloadFromCache(dir, mode, ctx); err == nil {
+				return nil
+			} else if !strings.Contains(err.Error(), "integrity check failed") {
+				log.Debug().Err(err).Msg("Cache download failed, trying original URL")
+			}
+		}
+	}
+
+	if err := r.downloadFromSource(dir, mode, ctx); err != nil {
+		return err
+	}
+
+	// Upload to cache if configured
+	if r.CacheUri != "" && os.Getenv("NO_CACHE_UPLOAD") == "" {
+		token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+		if token != "" {
+			if err := r.uploadToCache(dir, ctx); err != nil {
+				log.Debug().Err(err).Msg("Failed to upload to cache")
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Resource) downloadFromCache(dir string, mode os.FileMode, ctx context.Context) error {
+	token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+	if token == "" {
+		return fmt.Errorf("GRABIT_ARTIFACTORY_TOKEN must be set for cache operations")
```

Comment:
I don't think this is true.  See https://cardmaillouisville-my.sharepoint.com/:w:/g/personal/amdurh05_louisville_edu/ESdWGRZSS8xLuZ17je4MYrgBzV8s02ZNUPRyEtIMoGuc-A?rtime=TVqt2VMD3Ug